OpenSim/Simulation/Test/testStatesTrajectory.cpp

```diff
@@ -0,0 +1,877 @@
+    states.append(state); // 1
+    state.setTime(3.5);
+    states.append(state); // 2
+    state.setTime(4.1);
+    states.append(state); // 3
+    
+    // Nearest before.
+    SimTK_TEST(states.getIndexBefore(0.5) == 0);
+    SimTK_TEST(states.getIndexBefore(0.5000001) == 0);
+    SimTK_TEST(states.getIndexBefore(1.299999999) == 0);
+
+    SimTK_TEST(states.getIndexBefore(1.3) == 1);
+    SimTK_TEST(states.getIndexBefore(1.300001) == 1);
+    SimTK_TEST(states.getIndexBefore(3.49999) == 1);
+
+    SimTK_TEST(states.getIndexBefore(4.1) == 3);
+    SimTK_TEST(states.getIndexBefore(4.2) == 3);
+    SimTK_TEST(states.getIndexBefore(153.7) == 3);
+
+    // Nearest after.
+    SimTK_TEST(states.getIndexAfter(-100.0) == 0);
+    SimTK_TEST(states.getIndexAfter(0) == 0);
+    SimTK_TEST(states.getIndexAfter(0.1) == 0);
+    SimTK_TEST(states.getIndexAfter(0.4999999999) == 0);
+    SimTK_TEST(states.getIndexAfter(0.5) == 0);
+
+    SimTK_TEST(states.getIndexAfter(1.3000000000001) == 2);
+    SimTK_TEST(states.getIndexAfter(1.31) == 2);
+    SimTK_TEST(states.getIndexAfter(1.4) == 2);
+    SimTK_TEST(states.getIndexAfter(3.4) == 2);
+    SimTK_TEST(states.getIndexAfter(3.4999999999999) == 2);
+    SimTK_TEST(states.getIndexAfter(3.5) == 2);
+    
+    SimTK_TEST(states.getIndexAfter(4.0001) == 3);
+    SimTK_TEST(states.getIndexAfter(4.1) == 3);
+
+    // Exception if there is not a state before/after the specified time.
+    SimTK_TEST_MUST_THROW_EXC(states.getIndexBefore(0.25),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.getIndexBefore(0.0),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.getIndexBefore(-1.0),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.getIndexBefore(.499999),
+            StatesTrajectory::TimeOutOfRange);
+
+    SimTK_TEST_MUST_THROW_EXC(states.getIndexAfter(4.10000001),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.getIndexAfter(1050.9),
+            StatesTrajectory::TimeOutOfRange);
+
+    // Simulate the scenario that the tolerance is intended to handle.
+    {
+        // Pretend that the original states had times of 4.0 and 5.0, but
+        // that they were serialized (to a storage file) with limited precision
+        // and were written as 3.999999 and 5.000001. Now, the user
+        // thinks these times are still 4.0, and 5.0, and will end up
+        // getting the wrong states.
+        StatesTrajectory states2;
+        state.setTime(1.9);
+        states2.append(state); // 0
+        state.setTime(3.999999);
+        states2.append(state); // 1
+        state.setTime(4.5);
+        states2.append(state); // 2
+        state.setTime(5.000001);
+        states2.append(state); // 3
+        state.setTime(9.3);
+        states2.append(state); // 4
+        SimTK_TEST(states2.getIndexAfter(4.0) != 1);
+        SimTK_TEST(states2.getIndexBefore(5.0) != 3);
```

Comment:
My documentation of this test is a little confusing. This particular test is making sure that the methods give the *wrong* indices when an appropriate slop is *not* provided. That is, the user would expect to get 3, but didn't, because they did not provide the appropriate slop. I'm going to address this comment by improving the test's documentation. I'll also address the method names, but separately.